### PR TITLE
Improve documentation of callback functions

### LINKED ISSUE #153: Documentation should mention row-major or column-major flattening for jacobian.
The documentation specifies 

"The jacobian() and hessian() methods should return the non zero values as a falttened array."

but it is currently unclear whether this means row-major or column-major flattening.

### PR DESCRIPTION
**Pull request by chrhansk**
This should fix #153 and hopefully generally improve readability of the documentation of the callbacks.

### CODE REVIEW

**Comment by moorepants** [comment]
Thanks for this. I built it but it doesn't seem to work with the numpydoc format:

image

**Comment by moorepants** [comment]
I take that back, it does work, but the formatting is a bit unusual with the H1 or H2 in the sub bullet lists.

**Comment by moorepants** [comment]
Here is SciPy's solve_ivp docstring that also requires a callback function:

image

That seems sufficient to me and still fits in with the numpydoc format.

May I suggest that at least the "Parameters" and "Returns" headers in the sublist are turned into just bold `**Parameters**` or not have any formatting applied?

**Comment by chrhansk** [comment]
That is a good suggestion, I was not entirely happy with the formatting either. I made the changes you suggested, please have a look. 

This may be some issue with my setup, but it seems to me that part of the list of attributes is cut off and not visible in the documentation for some reason (i.e., the `hessianstructure` and `intermediate` sections). Can you confirm this?

File: cyipopt/cython/ipopt_wrapper.pyx [hunk 1]
```diff
@@ -160,63 +160,157 @@ cdef class Problem:
...
-                and column indices only). If ``None``, the Hessian is assumed
-                to be dense.
+                using `x`, `lambda` and `objective_factor`. The values should be
+                returned as a 1-dimensional array-like structure (using the same order as you
+                used when specifying the sparsity structure).
+
+                If this callback is not provided, the Hessian is calculated numerically.
+
+
+                **Parameters**
```
Comment (moorepants) [inline cyipopt/cython/ipopt_wrapper.pyx:277]: Is it somehow because of the multiple blank lines here?

Comment (moorepants) [reply cyipopt/cython/ipopt_wrapper.pyx:277]: That fixed it for me. Remove one blank line above `**Parameters**`.

Comment (moorepants) [reply cyipopt/cython/ipopt_wrapper.pyx:276]: I removed the line and pushed a commit.

Comment (chrhansk) [reply cyipopt/cython/ipopt_wrapper.pyx:276]: Thanks, appreciate it.

**Pull request merged**